Makefile

```diff
@@ -1,137 +1,205 @@
-	$(DOCKER_RUN_73) phpdbg -qrr bin/infection --threads=4
+.PHONY: test
+test:		## Runs all the tests
+# TODO: add a test to ensure we are not missing targets here
+test: test-unit test-e2e-phpdbg test-e2e-xdebug test-infection-phpdbg test-infection-xdebug
+
+.PHONY: test-unit
+test-unit:	## Runs the unit tests
+test-unit: test-unit-72 test-unit-73
 
+.PHONY: test-unit-72
+test-unit-72:	## Runs the unit tests for PHP 7.2
+test-unit-72: $(DOCKER_RUN_72_IMAGE) $(PHPUNIT)
+	$(DOCKER_RUN_72) $(PHPUNIT)
+
+.PHONY: test-unit-73
+test-unit-73:	## Runs the unit tests for PHP 7.3
+test-unit-73: $(DOCKER_RUN_73_IMAGE) $(PHPUNIT)
+	$(DOCKER_RUN_73) $(PHPUNIT)
 
-.PHONY: test-e2e-phpdbg test-e2e-phpdbg-72 test-e2e-phpdbg-73
-#e2e tests with phpdbg
+.PHONY: test-e2e-phpdbg
+test-e2e-phpdbg:	## Runs the end-to-end tests for PHPDBG
 test-e2e-phpdbg: test-e2e-phpdbg-72 test-e2e-phpdbg-73
 
-test-e2e-phpdbg-72: build-xdebug-72 $(INFECTION)
+.PHONY: test-e2e-phpdbg-72
+test-e2e-phpdbg-72:	## Runs the end-to-end tests for PHPDBG in PHP 7.2
+test-e2e-phpdbg-72: $(DOCKER_RUN_72_IMAGE) $(INFECTION)
 	$(DOCKER_RUN_72) env PHPDBG=1 ./tests/e2e_tests $(INFECTION)
 
-test-e2e-phpdbg-73: build-xdebug-73 $(INFECTION)
+.PHONY: test-e2e-phpdbg-73
+test-e2e-phpdbg-73:	## Runs the end-to-end tests for PHPDBG in PHP 7.3
+test-e2e-phpdbg-73: $(DOCKER_RUN_73_IMAGE) $(INFECTION)
 	$(DOCKER_RUN_73) env PHPDBG=1 ./tests/e2e_tests $(INFECTION)
 
+.PHONY: test-e2e-xdebug
+test-e2e-xdebug:	## Runs the end-to-end tests for xdebug
+test-e2e-xdebug: test-e2e-xdebug-72 test-e2e-xdebug-73
 
-.PHONY: test-infection-xdebug test-infection-xdebug-72 test-infection-xdebug-73
-#infection with xdebug
-test-infection-xdebug: test-infection-xdebug-72 test-infection-xdebug-73
+.PHONY: test-e2e-xdebug-72
+test-e2e-xdebug-72:	## Runs the end-to-end tests for xdebug in PHP 7.2
+test-e2e-xdebug-72: $(DOCKER_RUN_72_IMAGE) $(INFECTION)
+	$(DOCKER_RUN_72) ./tests/e2e_tests $(INFECTION)
 
-test-infection-xdebug-72: build-xdebug-72
-	$(DOCKER_RUN_72) php bin/infection --threads=4
+.PHONY: test-e2e-xdebug-73
+test-e2e-xdebug-73:	## Runs the end-to-end tests for xdebug in PHP 7.3
+test-e2e-xdebug-73: $(DOCKER_RUN_73_IMAGE) $(INFECTION)
+	$(DOCKER_RUN_73) ./tests/e2e_tests $(INFECTION)
 
-test-infection-xdebug-73: build-xdebug-73
-	$(DOCKER_RUN_73) php bin/infection --threads=4
+.PHONY: test-infection-phpdbg
+test-infection-phpdbg:		## Runs Infection with PHPDBG against itself
+test-infection-phpdbg: test-infection-phpdbg-72 test-infection-phpdbg-73
 
-.PHONY: test-e2e-xdebug test-e2e-xdebug-72 test-e2e-xdebug-73
-#e2e tests with xdebug
-test-e2e-xdebug: test-e2e-xdebug-72 test-e2e-xdebug-73
+.PHONY: test-infection-phpdbg-72
+test-infection-phpdbg-72:	## Runs Infection with PHPDBG against itself in PHP 7.2
+test-infection-phpdbg-72: $(DOCKER_RUN_72_IMAGE)
+	# TODO: calculate the number of thread here instead of using the arbitrary number 4
+	# TODO: check other occurrences
```

Comment:
noteworthy TODO

Reply:
More threads do not necessarily make things substantially better. My workstation has tens of threads, and I experimented to see if there's any benefit in running Infection over Infection using all available cores, and found out that anything more than 8 resulted in only marginal improvements, in my case.

Reply:
So you think it should stay at 4 instead of taking advantage of all the detected cores?

Reply:
I think four is fine, unless you want to look into this. Also it is a question how many processes Infection uses for a single mutant. I recall it being two for some valid reason, but can't look into the code right now. If so you'd have to divide number of your cores by two, and there's IPC which also can be a limiting factor.

Reply:
Let's leave it at that for now then...